### PR TITLE
board: arm: black_f407ve: Initial support

### PR DESCRIPTION
**Pull request by jukkar**
Add support for black_f407ve board which is based on stm32f407xx soc.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

The board is this one https://stm32-base.org/boards/STM32F407VET6-STM32-F4VE-V2.0.html

### CODE REVIEW

File: boards/arm/stm32_f4xx/Kconfig.board [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# STM32_F4XX board configuration
+
+# Copyright (c) 2016 Linaro Limited.
+# SPDX-License-Identifier: Apache-2.0
```
Comment (tbursztyka) [inline boards/arm/stm32_f4xx/Kconfig.board]: Add your © as well

File: boards/arm/stm32_f4xx/Kconfig.defconfig [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# STM32_F4XX board configuration
+
+# Copyright (c) 2016 Linaro Limited.
+# SPDX-License-Identifier: Apache-2.0
```
Comment (tbursztyka) [inline boards/arm/stm32_f4xx/Kconfig.defconfig]: same

File: boards/arm/stm32_f4xx/pinmux.c [hunk 3]
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2016 Linaro Limited.
```
Comment (tbursztyka) [inline boards/arm/stm32_f4xx/pinmux.c]: same

File: boards/arm/stm32_f4xx/stm32_f4xx.dts [hunk 4]
```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2017 Linaro Limited
```
Comment (tbursztyka) [inline boards/arm/stm32_f4xx/stm32_f4xx.dts]: same

File: boards/arm/stm32_f4xx/Kconfig.board (hunk 1 above)
Comment (jukkar) [reply boards/arm/stm32_f4xx/Kconfig.board]: Just existing files copied and only minor tweaking done here, I was thinking leaving the copyrights as is.

**Review by erwango — CHANGES_REQUESTED** [review]
Before proceeding, I've a complaint on the board name.
stm32_f4xx looks too generic,  vague and confusion prone (we have f4xx in soc, dts, and hal already)
Doing a few research, it seems this board is aka "Black F407VE", which I like much more and is used in other projects like STM32 Arduino Core.
Can you use this code name instead of stm32_f4xx ?

File: boards/arm/stm32_f4xx/Kconfig.board (hunk 1 above)
Comment (erwango) [reply boards/arm/stm32_f4xx/Kconfig.board]: On my side I'm pushing people that push boards to only put their own copyright on all files.
For sure this is always copy/paste of a previous board, but if we apply the same logic throughout the project life:
- Keeping the initial copyright, we end up with the same copyright for all boards files. But counting on the endless flow of small changes on this generic code the current files have nothing to do with the initial ones on which the copyright initially applied
- Keeping the initial copyright and adding your own, we end up with a list of copyright that reflect the copy/paste history but has no specific interest.

Using your copyright, we recognize the ownership of the one that did the effort of the porting (which include copy/paste but also validation and upstream effort).

**Comment by jukkar** [comment]
> stm32_f4xx looks too generic, vague and confusion prone (we have f4xx in soc, dts, and hal already)
> Doing a few research, it seems this board is aka "Black F407VE", which I like much more and is used in other projects like STM32 Arduino Core.
> Can you use this code name instead of stm32_f4xx ?

I used the name because the board has "stm32_f4ve" string in it. There are multiple versions of this board with slightly different soc so I named it "stm32_f4xx" (and that string can be found in some versions of the board).
For example
https://github.com/mcauser/BLACK_F407VE
https://github.com/mcauser/BLACK_F407ZE
https://github.com/mcauser/BLACK_F407ZG
Most probably the same board files could be used to cover all of them. So calling all of them with "black_f407ve" might be slightly confusing. One option could be to call it "black_f407xx"

**Comment by erwango** [comment]
> After some thought, perhaps we could name the board directory "black_f407xx" and inside it have board specific files like "black_f407ve.yaml", "black_f407ze", etc. What do you think?

I'm fine with using a generic board directory covering multiple targets. Similar approach have been used for stm32_min_dev. Though, I'm still not convinced of the *xx* suffix. What about *black_f407_generic*?

**Comment by jukkar** [comment]
> What about _black_f407_generic_?

Sounds good, lets go with that.

**Comment by jukkar** [comment]
Updated according to comments.

File: boards/arm/black_f407_generic/Kconfig.board [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# black_f407ve board configuration
+
+# Copyright (c) 2016 Linaro Limited.
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_BLACK_F407VE
+	bool "Black F407VE Development Board"
+	depends on SOC_STM32F407XG
```
Comment (erwango) [inline boards/arm/black_f407_generic/Kconfig.board]: Can you add the SOC_STM32F407XE variant (in a separate commit) ?

File: boards/arm/black_f407_generic/Kconfig.board [hunk 6]
```diff
@@ -0,0 +1,8 @@
+# black_f407ve board configuration
+
+# Copyright (c) 2016 Linaro Limited.
```
Comment (erwango) [inline boards/arm/black_f407_generic/Kconfig.board]: Use your own commit.

File: boards/arm/black_f407_generic/black_f407ve.dts [hunk 7]
```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2017 Linaro Limited
```
Comment (erwango) [inline boards/arm/black_f407_generic/black_f407ve.dts]: Use your own copyright

File: boards/arm/black_f407_generic/black_f407ve_defconfig [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+CONFIG_UART_CONSOLE=y
+
+# Enable pinmux
+CONFIG_PINMUX=y
+
+# Enable GPIO
+CONFIG_GPIO=y
+
+# Enable SPI
+CONFIG_SPI=y
```
Comment (erwango) [inline boards/arm/black_f407_generic/black_f407ve_defconfig]: Do not enable peripherals aside from essential ones: pinmux, clock, serial

File: boards/arm/black_f407_generic/black_f407ve_defconfig [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+CONFIG_CLOCK_STM32_PLL_M_DIVISOR=8
+CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER=336
+CONFIG_CLOCK_STM32_PLL_P_DIVISOR=2
+CONFIG_CLOCK_STM32_PLL_Q_DIVISOR=7
+CONFIG_CLOCK_STM32_AHB_PRESCALER=1
+CONFIG_CLOCK_STM32_APB1_PRESCALER=4
+CONFIG_CLOCK_STM32_APB2_PRESCALER=2
+
+# Enable pwm
+CONFIG_PWM=y
```
Comment (erwango) [inline boards/arm/black_f407_generic/black_f407ve_defconfig]: Disable this one too

File: boards/arm/black_f407_generic/pinmux.c [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+#include <pinmux/stm32/pinmux_stm32.h>
+
+#define STM32F4_PINMUX_FUNC_PC2_SPI2_MISO \
+	(STM32_PINMUX_ALT_FUNC_5 | STM32_PUPDR_PULL_DOWN)
+#define STM32F4_PINMUX_FUNC_PC3_SPI2_MOSI \
+	(STM32_PINMUX_ALT_FUNC_5 | STM32_PUPDR_PULL_DOWN)
+#define STM32F4_PINMUX_FUNC_PB10_SPI2_SCK \
+	(STM32_PINMUX_ALT_FUNC_5 | STM32_PUSHPULL_NOPULL | \
+	 STM32_OSPEEDR_VERY_HIGH_SPEED)
```
Comment (erwango) [inline boards/arm/black_f407_generic/pinmux.c]: These need to be define in appropriate f4 common file

**Review by erwango — CHANGES_REQUESTED** [review]

File: tests/drivers/spi/spi_loopback/boards/black_f407ve.conf [hunk 11]
```diff
@@ -0,0 +1,4 @@
+CONFIG_SPI=y
+CONFIG_SPI_LOOPBACK_DRV_NAME="SPI_2"
+CONFIG_SPI_STM32_INTERRUPT=y
+CONFIG_SPI_LOOPBACK_MODE_LOOP=y
```
Comment (erwango) [inline tests/drivers/spi/spi_loopback/boards/black_f407ve.conf]: I'm not familiar with this test, but shouldn't it  be done at project level?

File: tests/drivers/spi/spi_loopback/boards/black_f407ve.conf [hunk 12]
```diff
@@ -0,0 +1,4 @@
+CONFIG_SPI=y
```
Comment (erwango) [inline tests/drivers/spi/spi_loopback/boards/black_f407ve.conf]: To be done at project level

File: tests/drivers/spi/spi_loopback/boards/black_f407ve.conf [hunk 13]
```diff
@@ -0,0 +1,4 @@
...
+CONFIG_SPI_LOOPBACK_DRV_NAME="SPI_2"
+CONFIG_SPI_STM32_INTERRUPT=y
```
Comment (erwango) [inline tests/drivers/spi/spi_loopback/boards/black_f407ve.conf]: Done at board level.

**Review by erwango — CHANGES_REQUESTED** [review]
Minor comments on spi_loopback otherwise LGTM

File: tests/drivers/spi/spi_loopback/boards/black_f407ve.conf (hunk 11 above)
Comment (jukkar) [reply tests/drivers/spi/spi_loopback/boards/black_f407ve.conf]: Hmm, I am not sure I understand your point. I can remove the SPI loopback test, it was just there that I used to verify that the SPI was working correctly.

Comment (erwango) [reply tests/drivers/spi/spi_loopback/boards/black_f407ve.conf]: I meant the `CONFIG_SPI_LOOPBACK_MODE_LOOP=y` configuration.
Isn't it valid for all boards running the test  ?

Comment (jukkar) [reply tests/drivers/spi/spi_loopback/boards/black_f407ve.conf]: I removed the board from spi tests, it was just a leftover and should not have been there anyway.

**Review by erwango — APPROVED** [review]

**Pull request merged**